firmware/examples/FactoryTest/source/main.cpp

```diff
@@ -0,0 +1,323 @@
+           array[3]);
+  cs.SetHigh();
+  Delay(1000);
+  LOG_INFO("End of External Flash Test.");
+  if ((array[0] == 0x1F) && (array[1] == 0x40) && (array[2] == 0x00) &&
+      (array[3] == 0x00))
+  {
+    result = true;
+  }
+  return result;
+}
+
+bool I2cTest()
+{
+  // I2C device scan.
+  i2c.Initialize();
+  // Creates a status variable for I2C transactions
+  Status status;
+  int devices_found               = 0;
+  constexpr int total_i2c_devices = 3;
+  printf("++++++++++++++++++++++++++++++++++++++\n\n");
+  printf("Beginning I2c scan...\n\n");
+  for (uint8_t address = kFirstI2cAddress; address < kLastI2cAddress; address++)
+  {
+    status = i2c.Write(address, nullptr, 0);
+    if (status == Status::kSuccess)
+    {
+      LOG_INFO("    Found device at address: 0x%02X", address);
+      if (address == kAccelerometerAddress || address == kTemperatureAddress ||
+          address == kGestureAddress)
+      {
+        devices_found++;
+      }
+    }
+  }
+  bool i2c_pass           = (devices_found == total_i2c_devices);
+  bool gesture_pass       = GestureSensorTest();
+  bool accelerometer_pass = AccelerometerTest();
+  bool temp_pass          = TemperatureSensorTest();
+  bool final_result =
+      gesture_pass && accelerometer_pass && temp_pass && i2c_pass;
+  return final_result;
+}
+
+bool GpioFunctionalityTest()
+{
+  // P0_2 and P0_3 do not need to be tested.
+  // P0_29 and P0_30 do not need to be tested (lack of pullup/pulldown)
+  constexpr uint8_t num_ports = 5;
+  constexpr uint8_t num_pins  = 32;
+
+  const bool kLqfP80[num_ports][num_pins] = {
+    { 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1,
+      1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 0, 0 },
+    { 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 0, 0, 0, 1, 1,
+      0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1 },
+    { 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
+    { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0 },
+    { 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
+      0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 0, 0 }
+  };
+
+  printf("++++++++++++++++++++++++++++++++++++++\n\n");
+  printf("Begin Gpio test...\n\n");
+
+  uint8_t gpio_total                  = 0;
+  uint8_t gpio_passed                 = 0;
+  bool gpio_test[num_ports][num_pins] = { 0 };
+
+  for (uint8_t port = 0; port < num_ports; port++)
+  {
+    for (uint8_t pin = 0; pin < num_pins; pin++)
+    {
+      if (kLqfP80[port][pin])
+      {
+        Gpio test_pin(port, pin);
+        test_pin.SetAsOutput();
+        test_pin.SetLow();
+        if (test_pin.Read() == false)
+        {
+          gpio_test[port][pin] = true;
+        }
+        else
+        {
+          LOG_INFO("Pin[%u][%u] Value:%i", port, pin, test_pin.Read());
+          gpio_test[port][pin] = false;
+          if (kLqfP80[port][pin] == true && gpio_test[port][pin] == false)
```

Comment:
Why are you checking this again? This block of code can never be reached if `kLqfP80[port][pin] == false` due to the first `if` statement.

Reply:
`gpio_test[port][pin] == false` is always false, because the line directly above it sets it to false. Why is it being checked here?